### PR TITLE
Add UpdateTunnelConfiguration func

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

Implement UpdateTunnelConfiguration as described [here](https://api.cloudflare.com/#cloudflare-tunnel-configuration-put-configuration).

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] This change is using publicly documented (api.cloudflare.com or developers.cloudflare.com) and stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

### CODE REVIEW

File: tunnel.go [hunk 1]
```diff
@@ -211,6 +225,29 @@ func (api *API) UpdateTunnel(ctx context.Context, params TunnelUpdateParams) (Tu
...
 }
 
+// UpdateTunnelConfiguration updates an existing tunnel for the account.
+//
+// API reference: https://api.cloudflare.com/#cloudflare-tunnel-configuration-properties
+func (api *API) UpdateTunnelConfiguration(ctx context.Context, params TunnelConfigurationParams) (Tunnel, error) {
+	if params.AccountID == "" {
+		return Tunnel{}, ErrMissingAccountID
+	}
+
```
Comment (jacobbednarz) [inline tunnel.go:235]: you'll also want to check `TunnelID` is provided too

**Review by jacobbednarz — CHANGES_REQUESTED** [review]
thanks! we'll also need to add a unit test for this new endpoint

File: tunnel.go [hunk 2]
```diff
@@ -70,6 +77,87 @@ type TunnelUpdateParams struct {
...
+// OriginRequestConfig is a set of optional fields that users may set to
+// customize how cloudflared sends requests to origin services. It is used to set
+// up general config that apply to all rules, and also, specific per-rule
+// config.
+// Note: To specify a time.Duration in go-yaml, use e.g. "3s" or "24h".
+// This is copied directly from: https://github.com/cloudflare/cloudflared/blob/eed7d7bbc90ccaca129bc6b6c3fd18b19e6bf856/config/configuration.go#L189
+type OriginRequestConfig struct {
+	// HTTP proxy timeout for establishing a new connection
+	ConnectTimeout *time.Duration `yaml:"connectTimeout"`
```
Comment (jacobbednarz) [inline tunnel.go:95]: we aren't using YAML anywhere so all these tags can go. best stick to the publicly documented endpoint https://api.cloudflare.com/#cloudflare-tunnel-configuration-put-configuration

File: tunnel.go (hunk 1 above)
Comment (EngHabu) [reply tunnel.go:320]: Done

**Comment by EngHabu** [comment]
@jacobbednarz mind taking another look?

**Comment by jacobbednarz** [comment]
i ended up just fixing this myself as there were some foundational issues with this PR that would have made debugging difficult. 

- the structs had incorrect relationships. `TunnelConfiguration` has `OriginRequest` at the top level (not nested under `IngressRules`).
- the test wasn't testing the code path in use, it was only testing that the JSON was parsable which is was but not for the typed struct in use.
- the payload has stringified JSON instead of a typed response. updated both these methods to double unmarshal the payloads (first the v4 envelope then the `Config` after). i'm unsure why this decision was made but i'll look to see if we can change this as it doesn't make sense that we have a structured object stringified in a structured response.

**Comment by EngHabu** [comment]
Thank you @jacobbednarz

**Pull request merged**